src/integration-test/java/com/commercetools/sync/integration/externalsource/products/ProductSyncWithReferencedProductsIT.java

```diff
@@ -0,0 +1,470 @@
+
+        final ProductDraft productDraft = ProductDraftBuilder
+            .of(productType, ofEnglish("foo"), ofEnglish("foo-slug"), emptyList())
+            .key("foo")
+            .build();
+
+        final ProductDraft productDraft2 = ProductDraftBuilder
+            .of(productType, ofEnglish("foo2"), ofEnglish("foo-slug-2"), emptyList())
+            .key("foo2")
+            .build();
+
+        product = executeBlocking(CTP_TARGET_CLIENT.execute(ProductCreateCommand.of(productDraft)));
+        product2 = executeBlocking(CTP_TARGET_CLIENT.execute(ProductCreateCommand.of(productDraft2)));
+    }
+
+    private void clearSyncTestCollections() {
+        errorCallBackMessages = new ArrayList<>();
+        errorCallBackExceptions = new ArrayList<>();
+        warningCallBackMessages = new ArrayList<>();
+    }
+
+    private ProductSyncOptions buildSyncOptions() {
+        final Consumer<String> warningCallBack = warningMessage -> warningCallBackMessages.add(warningMessage);
+
+        return ProductSyncOptionsBuilder.of(CTP_TARGET_CLIENT)
+                                        .errorCallback(this::collectErrors)
+                                        .warningCallback(warningCallBack)
+                                        .build();
+    }
+
+    private void collectErrors(final String errorMessage, final Throwable exception) {
+        errorCallBackMessages.add(errorMessage);
+        errorCallBackExceptions.add(exception);
+    }
+
+    @AfterAll
+    static void tearDown() {
+        deleteProductSyncTestData(CTP_TARGET_CLIENT);
+    }
+
+    @Test
+    void sync_withProductReferenceAsAttribute_shouldCreateProductReferencingExistingProduct() {
+        // preparation
+        final AttributeDraft productReferenceAttribute =
+            AttributeDraft.of("product-reference", Reference.of(Product.referenceTypeId(), product.getKey()));
+        final ProductVariantDraft masterVariant = ProductVariantDraftBuilder
+            .of()
+            .sku("sku")
+            .key("new-product-master-variant")
+            .attributes(productReferenceAttribute)
+            .build();
+
+        final ProductDraft productDraftWithProductReference = ProductDraftBuilder
+            .of(productType, ofEnglish("productName"), ofEnglish("productSlug"), masterVariant)
+            .key("new-product")
+            .build();
+
+
+        // test
+        final ProductSync productSync = new ProductSync(syncOptions);
+        final ProductSyncStatistics syncStatistics =
+            productSync
+                .sync(singletonList(productDraftWithProductReference))
+                .toCompletableFuture()
+                .join();
+
+        // assertion
+        assertThat(syncStatistics).hasValues(1, 1, 0, 0);
+        assertThat(errorCallBackExceptions).isEmpty();
+        assertThat(errorCallBackMessages).isEmpty();
+        assertThat(warningCallBackMessages).isEmpty();
+
+
+        final Product createdProduct = CTP_TARGET_CLIENT
```

Comment:
maybe move duplicated assertion codes to a method

Reply:
In my opinion in this sense duplication is good. Because we want reading tests to be easy to read. (Imagine one test is failing. Debugging/reading it would be hard if we have to always jump to some refactored generalized method.)  It doesn't hurt us in this case duplication.

Reply:
I find this useful: https://stackoverflow.com/a/11837973/5654504

Reply:
Agree with this discussion did not know this term,  personally I am not checking the duplication on tests as you see there are lots of duplication other than that, but for this case, I thought it's simply an assertion, we don't need to interpret by ourself. 
(That checks the attributes are created on the platform too.) 

I also commented that it could be useful to see which update actions created during sync.
But of course, you could keep it if you think it's more readable.